test/e2e_node/device_plugin.go

```diff
@@ -0,0 +1,160 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Serial because the test restarts Kubelet
+var _ = framework.KubeDescribe("Device Plugin [Feature:DevicePlugin] [Serial] [Disruptive]", func() {
+	f := framework.NewDefaultFramework("device-plugin-errors")
+
+	Context("DevicePlugin", func() {
+		By("Enabling support for Device Plugin")
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.FeatureGates[string(features.DevicePlugins)] = true
+		})
+
+		BeforeEach(func() {
+
+		})
+
+		AfterEach(func() {
+
+		})
+
+		It("Checks that Kubelet handles device plugin registration correctly.", func() {
+
+			By("Start stub device plugin")
+			devs := []*pluginapi.Device{
+				{ID: "Dev-1", Health: pluginapi.Healthy},
+				{ID: "Dev-2", Health: pluginapi.Healthy},
+			}
+
+			dp1 := dp.NewDevicePluginStub(devs, pluginapi.DevicePluginPath+"dp.1")
+			err := dp1.Start()
+			framework.ExpectNoError(err)
+
+			By("Wait for node is ready")
+			Eventually(func() int {
+				nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+				return len(nodeList.Items)
+			}, 30*time.Second, framework.Poll).Should(Equal(1))
+
+			By("Register resources")
+			resourceName := "mock1.com/resource"
+			err = dp1.Register(pluginapi.KubeletSocket, resourceName)
+			framework.ExpectNoError(err)
+
+			By("Waiting for stub device plugin to become available on the local node")
+			devsLen := int64(len(devs))
+			Eventually(func() int64 {
+				return framework.NumberOfDevices(getLocalNode(f), resourceName)
+			}, 30*time.Second, framework.Poll).Should(Equal(devsLen))
+
+			By("Creating one pod on node with at least one fake-device")
+			podRECMD := "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs"
+			pod1 := f.PodClient().CreateSync(framework.MakePauseImage(busyboxImage, resourceName, podRECMD))
+			deviceIDRE := "stub devices: (Dev-[0-9]+)"
+			count1, devId1 := framework.GetDeviceIDWithRE(f, pod1.Name, pod1.Name, 0, deviceIDRE)
+
+			Expect(devId1).To(Not(Equal("")))
+			pod1, err = f.PodClient().Get(pod1.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+			By("Restarting Kubelet and waiting for the current running pod to restart")
+			framework.RestartKubeletForDP()
+
+			By("Wait for node is ready")
+			Eventually(func() int {
+				nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+				return len(nodeList.Items)
+			}, 30*time.Second, framework.Poll).Should(Equal(1))
+
+			By("Confirming that after a kubelet and pod restart, fake-device assignement is kept")
+			count1, devIdRestart1 := framework.GetDeviceIDWithRE(f, pod1.Name, pod1.Name, count1+1, deviceIDRE)
+			Expect(devIdRestart1).To(Equal(devId1))
+
+			By("Re-Register resources")
+			dp1 = dp.NewDevicePluginStub(devs, pluginapi.DevicePluginPath+"dp.1")
+			err = dp1.Start()
+			framework.ExpectNoError(err)
+
+			err = dp1.Register(pluginapi.KubeletSocket, resourceName)
```

Comment:
Any reason we have to call NewDevicePluginStub() again?

Reply:
re-register will fail without `NewDevicePluginStub()`

Reply:
I try to Stop() -> Start() -> Register(), but there is something wrong: `listAndWatch ended unexpectedly for device plugin fake.com/resource with error EOF`.
I will take a look at this later, then fix this in another PR :)